src/environments/ctmrgenv.jl

```diff
@@ -8,44 +8,272 @@ struct CTMRGEnv{C,T}
     edges::Array{T,3}
 end
 
+_spacetype(::Int) = ComplexSpace
+_spacetype(::S) where {S<:ElementarySpace} = S
+
+_to_space(χ::Int) = ℂ^χ
+_to_space(χ::ElementarySpace) = χ
```

Comment:
I like this approach, is there any chance you can also do something like this for the `InfinitePEPS` constructors?

Reply:
I don't think this would help with the `InfinitePEPS` constructors themselves, since these never have to make a distinction but just delegate to the `PEPSTensor` constructor. For the latter, I think it's a bit tricky to squash the two types of constructors since it's impossible to specify a dual space when passing only integer dimensions.

Specifically, it seems to me this
https://github.com/QuantumKitHub/PEPSKit.jl/blob/5cd0304a0475ff880d7ceb0f534f1fc90bf9b842/src/states/abstractpeps.jl#L39
choice of automatically dualizing the south and west spaces makes it hard to merge the two cases .

Reply:
Ok, I didn't really look into it in detail, I just had a vague memory of a similar pattern.